include/cpp2util.h

```diff
@@ -494,6 +494,101 @@ class out {
+    template <typename X > friend void operator- (X, const safetychecked_pointer&)  {static_assert(program_violates_lifetime_safety_guarantee<X>, "pointer arithmetic is illegal - use std::span or gsl::span instead");}
+    template <typename X > friend void operator* (X, const safetychecked_pointer&)  {static_assert(program_violates_lifetime_safety_guarantee<X>, "pointer arithmetic is illegal - use std::span or gsl::span instead");}
+    template <typename X > friend void operator/ (X, const safetychecked_pointer&)  {static_assert(program_violates_lifetime_safety_guarantee<X>, "pointer arithmetic is illegal - use std::span or gsl::span instead");}
+    template <typename X > friend void operator% (X, const safetychecked_pointer&)  {static_assert(program_violates_lifetime_safety_guarantee<X>, "pointer bitwise manipulation is illegal - use std::bit_cast to convert to raw bytes first");}
+    template <typename X > friend void operator^ (X, const safetychecked_pointer&)  {static_assert(program_violates_lifetime_safety_guarantee<X>, "pointer bitwise manipulation is illegal - use std::bit_cast to convert to raw bytes first");}
+    template <typename X > friend void operator& (X, const safetychecked_pointer&)  {static_assert(program_violates_lifetime_safety_guarantee<X>, "pointer bitwise manipulation is illegal - use std::bit_cast to convert to raw bytes first");}
+    template <typename X > friend void operator| (X, const safetychecked_pointer&)  {static_assert(program_violates_lifetime_safety_guarantee<X>, "pointer bitwise manipulation is illegal - use std::bit_cast to convert to raw bytes first");}
+
+
+    template <typename X>
+        requires (std::is_same_v<T,X> || std::is_base_of_v<T, X>)
+    constexpr safetychecked_pointer& operator=(X lhs) noexcept {
+        ptr = lhs;
+        return *this;
+    }
+
+    template <typename X>
+        requires std::is_same_v<std::nullptr_t,X>
+    constexpr void operator=(X lhs) noexcept { static_assert(program_violates_lifetime_safety_guarantee<X>, "pointer assignment from null is illegal"); }
+
+    template <typename X>
+        requires std::is_integral_v<X>
+    constexpr void operator=(X lhs) noexcept { static_assert(program_violates_lifetime_safety_guarantee<X>, "pointer assignment from integer is illegal"); }
+
+    bool operator!() const { return !ptr; }
+
+    constexpr safetychecked_pointer<T*> operator&() noexcept { return &ptr; }
+
+    constexpr auto operator*() noexcept {
+        if constexpr (std::is_pointer_v<CPP2_TYPEOF(*ptr)>) {
+            return safetychecked_pointer<CPP2_TYPEOF(*ptr)>(*ptr);
+        } else {
+            return *ptr;
+        }
+    }
+
+    constexpr T operator->() const noexcept { return ptr; }
+};
+
+#define CPP2_SAFETYCHECK(x) []<typename T>(T&& o) { \
+    if constexpr (std::is_pointer_v<std::remove_cvref_t<T>>) { \
+        return cpp2::safetychecked_pointer(o); \
+    } else { \
+        return std::forward<T>(o); \
+    } \
+}(x)
```

Comment:
I think cppfront does this elsewhere. But since you're using a macro anyways, I think it'd be better to rely on guaranteed copy elision and return `x` directly. There are types for which this current implementation will have bad performance, like big aggregates or those for which move is copy.

```suggestion
#define CPP2_SAFETYCHECK(x) []() { \
    if constexpr (std::is_pointer_v<std::remove_cvref_t<decltype(x)>>) { \
        return cpp2::safetychecked_pointer(x); \
    } else { \
        return x; \
    } \
}()
```

Reply:
Ah, yes, I forgot `[&]`. Would that suffice?

Reply:
Actually, that has problems with bit-fields, and structured bindings in C++20.

Reply:
> What I did was not use moving but perfect forwarding - that is why the lambda is a template.

Yes, but that'll compile to a copy for big PODs. Actually a move, but for big PODs, there's no difference.

Reply:
Ok. I will test that. If you can think about test cases I can verify them.

Reply:
I think something like this could work without tripping due to big PODs, bit-fields and structured bindings: https://godbolt.org/z/bovP3vfvo.

```C++
template<class U>
using safetychecked_pointer_impl = decltype([]<typename T>(T&& o) {
    if constexpr (std::is_pointer_v<T>) {
        return safetychecked_pointer(o);
    } else {
        return std::forward<T>(o);
    }
}(std::declval<U>()));

#define CPP2_SAFETYCHECK(x) safetychecked_pointer_impl<decltype((x))>(x)
```

Reply:
This macro will be used only in declaring variables - to help deduce the type and introduce safety checks. So probably there will be no issue with coping as there will be coping in the end. But current implementation still has issues with assignments to reference.

```cpp
std::array<double, 64> tmp {1,2,3,4,5,6,7,8,9,0};

void fun_ref() {
    auto& a = CPP2_SAFETYCHECK(tmp);
}
```
will end up with an error:
```
error: cannot bind non-const lvalue reference of type 'std::array<double, 64>&' to an rvalue of type 'std::array<double, 64>'
   90 | #define CPP2_SAFETYCHECK(x) []<typename T>(T&& o) { \
      |                             ~~~~~~~~~~~~~~~~~~~~~~~~~
   91 |     if constexpr (std::is_pointer_v<T>) { \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   92 |         return safetychecked_pointer(o); \
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   93 |     } else { \
      |     ~~~~~~~~~~
   94 |         return std::forward<T>(o); \
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   95 |     } \
      |     ~~~
   96 | }(x)
      | ~^~~
<source>:101:15: note: in expansion of macro 'CPP2_SAFETYCHECK'
  101 |     auto& a = CPP2_SAFETYCHECK(tmp);
```

That can be solved by using `decltype(auto)`:
```cpp
#define CPP2_SAFETYCHECK(x) []<typename T>(T&& o) -> decltype(auto) { \
    if constexpr (std::is_pointer_v<T>) { \
        return safetychecked_pointer(o); \
    } else { \
        return std::forward<T>(o); \
    } \
}(x)
```

The working prototype is here: https://godbolt.org/z/reTqKs18h

I will check how it works with the other types you mentioned.

Reply:
There _might_ be no copying in some cases, but when `x` is a prvalue, you've initialized `o` from `x`, before `o` is used to initialize the user-declared variable. I think that'll fail for non-movable non-copyable types. My suggestion to fold the return type of the lambda should be able to make the user-declared variable be initialized directly from `x` when the type of `x` is not a pointer.

Reply:
OK, thank you for your feedback. I decided to remove the macro entirely and switch to the `is()`/`as()` functions style.

Instead of the `CPP2_SAFETYCHECK` macro now there are `cpp2::safety_check()` functions with overloads that check requirements for types. Probably we can tweak the performance. My goal for now is to make it working for all cases.

I have tested the implementation with bitfields and unique_ptr:
```cpp
auto up = std::make_unique<int>(42);

void fun_noncopyable() {
    auto a = safety_check(up);
    auto b = safety_check(std::make_unique<int>(44));
}

struct S {
    unsigned int b : 3;
} s{ 3 };

void fun_bitfields() {
    auto a = safety_check(s.b);
}
```

Previous cases still work:
```cpp
std::array<double, 64> tmp {1,2,3,4,5,6,7,8,9,0};

void fun_ref() {
    auto& a = safety_check(tmp);
}

void fun_copy() {
    auto a = safety_check(tmp);
}

void fun_rvalue() {
    auto a = safety_check(std::array<double, 64>{9,8,7,6,5,4,3,2,1});
}

void fun_ptr() {
    auto a = safety_check(&tmp);
}
```

Working prototype: https://godbolt.org/z/4n5Yraend